File: src/main/java/cz/stasimek/fakturaceeasypeasy/util/InvoiceNumberGenerator.java
package cz.stasimek.fakturaceeasypeasy.util;

import cz.stasimek.fakturaceeasypeasy.entity.Invoice;

public class InvoiceNumberGenerator {

	private InvoiceNumberGenerator() {
	}

	public static String generate(String lastInvoiceNumber, int currentYear) {
		// First invoice at all
		if (lastInvoiceNumber == null) {
			return currentYear + "0001";
		}

		// First invoice this year
		if (!lastInvoiceNumber.startsWith(currentYear + "")) {
			if (lastInvoiceNumber.contains("-")) {
				return currentYear + "-0001";
			} else {
				return currentYear + "0001";
			}
		}

		// Non-first invoice this year
		String[] numberParts = lastInvoiceNumber.split("-");
		if (numberParts.length == 1) {
			return Long.parseLong(lastInvoiceNumber) + 1 + "";
		} else {
			return currentYear + "-" + String.format("%04d", Long.parseLong(numberParts[1]) + 1);
		}
	}
}

File: src/main/kotlin/cz/stasimek/fakturaceeasypeasy/util/InvoiceNumberGenerator.kt
package cz.stasimek.fakturaceeasypeasy.util

object InvoiceNumberGenerator {

    fun generate(lastInvoiceNumber: String?, currentYear: Int): String {
        // First invoice at all
        if (lastInvoiceNumber == null) {
            return "${currentYear}0001"
        }

        // First invoice this year
        if (!lastInvoiceNumber.startsWith(currentYear.toString())) {
            if (lastInvoiceNumber.contains("-")) {
                return "$currentYear-0001"
            } else {
                return "${currentYear}0001"
            }
        }

        // Non-first invoice this year
        val numberParts = lastInvoiceNumber.split("-")
        if (numberParts.size == 1) {
            return (lastInvoiceNumber.toLong() + 1).toString()
        } else {
            return "$currentYear-" + String.format("%04d", numberParts[1].toLong() + 1)
        }
    }
}